### DISCUSSION TITLE
Forcing async BackgroundTasks to be run as if it were sync under the hood with run_in_threadpool()

### DISCUSSION THREAD

**Discussion by LuchiLucs** [fastapi/fastapi#12913] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
app = FastAPI()


async def async_task():
    import asyncio
    import time
    await asyncio.sleep(1)  # simulate sequential non-blocking code
    time.sleep(2)           # simulate sequential blocking code
    await asyncio.sleep(3)  # simulate sequential non-blocking code
    time.sleep(4)           # simulate sequential blocking code

@app.get("/")
async def run_task(tasks: BackgroundTasks):
    tasks.add_task(async_task)
```

### Description

I wish I could run `async_task` as if it was not `async` so that under the hood is not run in the same process as fastapi blocking the event loop, but run concurrenlty in another thread by means of `run_in_threadpool()`. How can I do it?

### Operating System

Linux

### FastAPI Version

0.115.0

### Pydantic Version

2.9.2

### Python Version

3.12.6

**Comment by YuriiMotov** [comment, marked as answer]
`async` functions should be run in event loop.
`run_in_threadpool` is for `sync` functions.

What you probably want is to run this task in a separate event loop.
There is no way to configure it in FastAPI.

As a solution, you can create a function that will accept coroutine as a parameter, create new event loop (in a separate thread), run target function in created event loop and await it.
This way it will not block your main event loop even if your task will contain blocking code.

(code created with ChatGPT and haven't been tested properly, but seems to be working):

```python
import asyncio
import threading
from typing import Any, Coroutine

from fastapi import FastAPI, BackgroundTasks


def run_coroutine_in_new_loop_async(coro: Coroutine[Any, Any, Any]):
    async def _run_coroutine_in_new_loop_async() -> Any:
        loop = asyncio.get_running_loop()
        future: asyncio.Future = loop.create_future()

        def thread_worker():
            try:
                thread_loop = asyncio.new_event_loop()
                asyncio.set_event_loop(thread_loop)

                result = thread_loop.run_until_complete(coro)
                thread_loop.call_soon_threadsafe(thread_loop.stop)
                thread_loop.close()

                loop.call_soon_threadsafe(future.set_result, result)
            except Exception as e:
                loop.call_soon_threadsafe(future.set_exception, e)

        threading.Thread(target=thread_worker, daemon=True).start()
        return await future

    return _run_coroutine_in_new_loop_async


app = FastAPI()


async def async_task():
    import time
    print("Starting async task")
    time.sleep(10)           # simulate sequential blocking code
    print("Async task finished")

@app.get("/")
async def run_task(tasks: BackgroundTasks):
    tasks.add_task(
        run_coroutine_in_new_loop_async(async_task())
    )

```

Run and perform several requests.
Results:

```
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     127.0.0.1:47790 - "GET / HTTP/1.1" 200 OK
Starting async task
INFO:     127.0.0.1:47790 - "GET / HTTP/1.1" 200 OK
Starting async task
INFO:     127.0.0.1:47790 - "GET / HTTP/1.1" 200 OK
Starting async task
Async task finished
Async task finished
Async task finished
```

Also, consider using distributed task queues like Celery, ARQ, etc..
